scrabbler/simulation.py

```diff
@@ -0,0 +1,127 @@
+               "RRRRRRSSSS"
+               "TTTTTTUUUU"
+               "VVWWXYYZ??")
+    
+    RACK_SIZE = 7
+
+    @staticmethod
+    def simulate_game():
+        Simulation().simulate()
+
+    def __init__(self):
+        # Initialize game and board
+        self.game = Game()
+        self.board = self.game.board
+        # Rack starts out empty.
+        self.racks = [[], []]
+        self.scores = [0, 0]
+        # List of letters we can still pick from.
+        self.bag = list(Simulation.LETTERS)
+        self.player = 0
+        self.endgame = False
+        self.times = []
+
+    def simulate(self):
+        # Keep playing until we're out of tiles or solutions.
+        while self.exectute_turn():
+            # switch whose turn it is
+            self.player = 1 - self.player
+            # Show the game board. We could have also done print(board) here.
+            self.game.show()
+
+        self.print_end_game_message()
+
+    def exectute_turn(self):
+        self.generate_new_rack()
+
+        # End game once either player has no letters left
+        if not self.racks[self.player]:
+            return False
+
+        best_move = self.find_best_move()
+
+        # If a valid move exists, then make best move, otherwise end game
+        if best_move:
+            self.make_move(best_move)
+            return True
+        else:
+            return False
+
+    def generate_new_rack(self):
+        print("########################## Player %d turn ############################"%(self.player + 1))
+        print("Bag: %s" % "".join(self.bag))
+        print("Player %d rack pre-draw: %s" % (self.player + 1, self.racks[self.player]))
+        self.generate_rack_and_bag()
+        print("Player %d rack post-draw: %s" % (self.player + 1, self.racks[self.player]))
+
+    def generate_rack_and_bag(self):
+        """Randomly chooses tiles from bag and places in rack"""
+        for i in range(Simulation.RACK_SIZE - len(self.racks[self.player])):
+            # If bag has ended then end game begins (as of right now this 
+            # doesn't formally mean anything, just print statement)
+            if not self.bag:
+                if not self.endgame:
+                    print('|||||||||||||||||||| END GAME STARTS NOW ||||||||||||||||||||')
+                    self.endgame = True
+                break
+
+            new_tile = random.choice(self.bag)
+            self.racks[self.player].append(new_tile)
+            self.bag.remove(new_tile)
+
+    def find_best_move(self):
+        # This function simply finds all valid moves and returns them ordered by score. The second parameter, which
+        # is unused right now, used to limit the number of moves it returned, but I changed it so that all moves are
+        # returned
+        before = time.time()
+        best_moves = self.game.find_best_moves(''.join(self.racks[self.player]))
+        self.times.append(time.time() - before)
+        return best_moves[0]
```

Comment:
best_moves might be empty since there might be no valid moves, so this should check if it is empty and return None if so.

Reply:
Good catch, I'll fix it.

Reply:
Since self.game.find_best_moves() returns all valid moves, we might want to consider changing the name to something like generate_all_valid_moves() since all of our algorithms will require a function to generate all valid moves.

Reply:
I've renamed the function `find_valid_moves` and removed the unused optional parameter `num`.